presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkOutputOperator.java

```diff
@@ -0,0 +1,295 @@
+    }
+
+    @Override
+    public OperatorContext getOperatorContext()
+    {
+        return operatorContext;
+    }
+
+    @Override
+    public ListenableFuture<?> isBlocked()
+    {
+        return rowBuffer.isFull();
+    }
+
+    @Override
+    public boolean needsInput()
+    {
+        return !finished && isBlocked().isDone();
+    }
+
+    @Override
+    public void addInput(Page page)
+    {
+        page = pagePreprocessor.apply(page);
+        int positionCount = page.getPositionCount();
+        int channelCount = page.getChannelCount();
+        int averageRowSizeInBytes = min(toIntExact(page.getLogicalSizeInBytes() / positionCount), 10);
+        Page partitionFunctionArguments = getPartitionFunctionArguments(page);
+        for (int position = 0; position < positionCount; position++) {
+            SliceOutput output = new DynamicSliceOutput(averageRowSizeInBytes * 2);
+            for (int channel = 0; channel < channelCount; channel++) {
+                Block block = page.getBlock(channel);
+                block.writePositionTo(position, output);
+            }
+
+            boolean shouldReplicate = (replicateNullsAndAny && !hasAnyRowBeenReplicated) ||
+                    nullChannel.isPresent() && page.getBlock(nullChannel.getAsInt()).isNull(position);
+            if (shouldReplicate) {
+                for (int i = 0; i < partitionFunction.getPartitionCount(); i++) {
+                    rowBuffer.enqueue(new PrestoSparkRow(i, output.size(), output.getUnderlyingSlice().byteArray()));
+                }
+                hasAnyRowBeenReplicated = true;
+            }
+            else {
+                int partition = getPartition(partitionFunctionArguments, position);
+                rowBuffer.enqueue(new PrestoSparkRow(partition, output.size(), output.getUnderlyingSlice().byteArray()));
+            }
+        }
+    }
+
+    private int getPartition(Page partitionFunctionArgs, int position)
+    {
+        return partitionFunction.getPartition(partitionFunctionArgs, position);
+    }
+
+    private Page getPartitionFunctionArguments(Page page)
+    {
+        Block[] blocks = new Block[partitionChannels.size()];
+        for (int i = 0; i < blocks.length; i++) {
+            Optional<Block> partitionConstant = partitionConstants.get(i);
+            if (partitionConstant.isPresent()) {
+                blocks[i] = new RunLengthEncodedBlock(partitionConstant.get(), page.getPositionCount());
+            }
+            else {
+                blocks[i] = page.getBlock(partitionChannels.get(i));
+            }
+        }
+        return new Page(page.getPositionCount(), blocks);
+    }
+
+    @Override
+    public Page getOutput()
+    {
+        return null;
+    }
+
+    @Override
+    public void finish()
+    {
+        finished = true;
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return finished && isBlocked().isDone();
```

Comment:
is this enough -- do we want to make sure `rowBuffer` is empty?

Reply:
It looks like the `TaskOutputFactory` is not waiting for the buffer to be drained. I followed the same approach here.